// @tanstack/ai/dist/esm/activities/summarize/index.js
import { aiEventClient } from "../../event-client.js";
const kind = "summarize";
function createId(prefix) {
  return `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
}
function summarize(options) {
  const { stream } = options;
  if (stream) {
    return runStreamingSummarize(
      options
    );
  }
  return runSummarize(
    options
  );
}
async function runSummarize(options) {
  const { adapter, text, maxLength, style, focus } = options;
  const model = adapter.model;
  const requestId = createId("summarize");
  const inputLength = text.length;
  const startTime = Date.now();
  aiEventClient.emit("summarize:request:started", {
    requestId,
    provider: adapter.name,
    model,
    inputLength,
    timestamp: startTime
  });
  const summarizeOptions = {
    model,
    text,
    maxLength,
    style,
    focus
  };
  const result = await adapter.summarize(summarizeOptions);
  const duration = Date.now() - startTime;
  const outputLength = result.summary.length;
  aiEventClient.emit("summarize:request:completed", {
    requestId,
    provider: adapter.name,
    model,
    inputLength,
    outputLength,
    duration,
    timestamp: Date.now()
  });
  return result;
}
async function* runStreamingSummarize(options) {
  const { adapter, text, maxLength, style, focus } = options;
  const model = adapter.model;
  const summarizeOptions = {
    model,
    text,
    maxLength,
    style,
    focus
  };
  if (adapter.summarizeStream) {
    yield* adapter.summarizeStream(summarizeOptions);
    return;
  }
  const result = await adapter.summarize(summarizeOptions);
  yield {
    type: "TEXT_MESSAGE_CONTENT",
    messageId: result.id,
    model: result.model,
    timestamp: Date.now(),
    delta: result.summary,
    content: result.summary
  };
  yield {
    type: "RUN_FINISHED",
    runId: result.id,
    model: result.model,
    timestamp: Date.now(),
    finishReason: "stop",
    usage: result.usage
  };
}
function createSummarizeOptions(options) {
  return options;
}
export {
  createSummarizeOptions,
  kind,
  summarize
};
