// @tanstack/ai/dist/esm/activities/generateSpeech/index.js
import { aiEventClient } from "../../event-client.js";
const kind = "tts";
function createId(prefix) {
  return `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
}
async function generateSpeech(options) {
  const { adapter, ...rest } = options;
  const model = adapter.model;
  const requestId = createId("speech");
  const startTime = Date.now();
  aiEventClient.emit("speech:request:started", {
    requestId,
    provider: adapter.name,
    model,
    text: rest.text,
    voice: rest.voice,
    format: rest.format,
    speed: rest.speed,
    modelOptions: rest.modelOptions,
    timestamp: startTime
  });
  return adapter.generateSpeech({ ...rest, model }).then((result) => {
    const duration = Date.now() - startTime;
    aiEventClient.emit("speech:request:completed", {
      requestId,
      provider: adapter.name,
      model,
      audio: result.audio,
      format: result.format,
      audioDuration: result.duration,
      contentType: result.contentType,
      duration,
      modelOptions: rest.modelOptions,
      timestamp: Date.now()
    });
    return result;
  });
}
function createSpeechOptions(options) {
  return options;
}
export {
  createSpeechOptions,
  generateSpeech,
  kind
};
