// @tanstack/ai-client/dist/esm/events.js
import { aiEventClient } from "@tanstack/ai/event-client";
class ChatClientEventEmitter {
  constructor(clientId) {
    this.clientId = clientId;
  }
  /**
   * Emit client created event
   */
  clientCreated(initialMessageCount) {
    this.emitEvent("client:created", {
      initialMessageCount
    });
  }
  /**
   * Emit loading state changed event
   */
  loadingChanged(isLoading) {
    this.emitEvent("client:loading:changed", { isLoading });
  }
  /**
   * Emit error state changed event
   */
  errorChanged(error) {
    this.emitEvent("client:error:changed", {
      error
    });
  }
  /**
   * Emit text update events (combines processor and client events)
   */
  textUpdated(streamId, messageId, content) {
    this.emitEvent("text:chunk:content", {
      streamId,
      messageId,
      content
    });
  }
  /**
   * Emit tool call state change events (combines processor and client events)
   */
  toolCallStateChanged(streamId, messageId, toolCallId, toolName, state, args) {
    this.emitEvent("tools:call:updated", {
      streamId,
      messageId,
      toolCallId,
      toolName,
      state,
      arguments: args
    });
  }
  /**
   * Emit tool result state change event
   */
  /**
   * Emit thinking update event
   */
  thinkingUpdated(streamId, messageId, content, delta) {
    this.emitEvent("text:chunk:thinking", {
      streamId,
      messageId,
      content,
      delta
    });
  }
  /**
   * Emit approval requested event
   */
  approvalRequested(streamId, messageId, toolCallId, toolName, input, approvalId) {
    this.emitEvent("tools:approval:requested", {
      streamId,
      messageId,
      toolCallId,
      toolName,
      input,
      approvalId
    });
  }
  /**
   * Emit message appended event
   */
  messageAppended(uiMessage, streamId) {
    const content = uiMessage.parts.filter((part) => part.type === "text").map((part) => part.content).join(" ");
    this.emitEvent("text:message:created", {
      streamId,
      messageId: uiMessage.id,
      role: uiMessage.role,
      content,
      parts: uiMessage.parts
    });
  }
  /**
   * Emit message sent event.
   * Supports both simple string content and multimodal content arrays.
   *
   * @param messageId - The ID of the sent message
   * @param content - The message content (string or array of ContentPart for multimodal)
   */
  messageSent(messageId, content) {
    const textContent = typeof content === "string" ? content : content.filter((part) => part.type === "text").map((part) => part.content).join("");
    this.emitEvent("text:message:created", {
      messageId,
      role: "user",
      content: textContent,
      // Include full content for multimodal messages
      ...Array.isArray(content) && { parts: content }
    });
    this.emitEvent("text:message:user", {
      messageId,
      role: "user",
      content: textContent,
      // Include full content for multimodal messages
      ...Array.isArray(content) && { parts: content }
    });
  }
  /**
   * Emit reloaded event
   */
  reloaded(fromMessageIndex) {
    this.emitEvent("client:reloaded", {
      fromMessageIndex
    });
  }
  /**
   * Emit stopped event
   */
  stopped() {
    this.emitEvent("client:stopped");
  }
  /**
   * Emit messages cleared event
   */
  messagesCleared() {
    this.emitEvent("client:messages:cleared");
  }
  /**
   * Emit tool result added event
   */
  toolResultAdded(toolCallId, toolName, output, state) {
    this.emitEvent("tools:result:added", {
      toolCallId,
      toolName,
      output,
      state
    });
  }
  /**
   * Emit tool approval responded event
   */
  toolApprovalResponded(approvalId, toolCallId, approved) {
    this.emitEvent("tools:approval:responded", {
      approvalId,
      toolCallId,
      approved
    });
  }
}
class DefaultChatClientEventEmitter extends ChatClientEventEmitter {
  /**
   * Emit an event with automatic clientId and timestamp for client/tool events
   */
  emitEvent(eventName, data) {
    if (eventName.startsWith("client:") || eventName.startsWith("tools:") || eventName.startsWith("text:")) {
      aiEventClient.emit(eventName, {
        ...data,
        clientId: this.clientId,
        source: "client",
        timestamp: Date.now()
      });
    } else {
      aiEventClient.emit(eventName, {
        ...data,
        timestamp: Date.now()
      });
    }
  }
}
export {
  ChatClientEventEmitter,
  DefaultChatClientEventEmitter
};
