// @tanstack/ai-client/dist/esm/index.js
import { ChatClient } from "./chat-client.js";
import { clientTools, createChatClientOptions } from "./types.js";
import { fetchHttpStream, fetchServerSentEvents, rpcStream, stream } from "./connection-adapters.js";
import { BatchStrategy, CompositeStrategy, ImmediateStrategy, PartialJSONParser, PunctuationStrategy, StreamProcessor, WordBoundaryStrategy, convertMessagesToModelMessages, defaultJSONParser, generateMessageId, modelMessageToUIMessage, modelMessagesToUIMessages, normalizeToUIMessage, parsePartialJSON, uiMessageToModelMessages } from "@tanstack/ai";
export {
  BatchStrategy,
  ChatClient,
  CompositeStrategy,
  ImmediateStrategy,
  PartialJSONParser,
  PunctuationStrategy,
  StreamProcessor,
  WordBoundaryStrategy,
  clientTools,
  convertMessagesToModelMessages,
  createChatClientOptions,
  defaultJSONParser,
  fetchHttpStream,
  fetchServerSentEvents,
  generateMessageId,
  modelMessageToUIMessage,
  modelMessagesToUIMessages,
  normalizeToUIMessage,
  parsePartialJSON,
  rpcStream,
  stream,
  uiMessageToModelMessages
};
