// @tanstack/ai/dist/esm/activities/generateVideo/index.js
import { aiEventClient } from "../../event-client.js";
const kind = "video";
function createId(prefix) {
  return `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
}
async function generateVideo(options) {
  const { adapter, prompt, size, duration, modelOptions } = options;
  const model = adapter.model;
  return adapter.createVideoJob({
    model,
    prompt,
    size,
    duration,
    modelOptions
  });
}
async function getVideoJobStatus(options) {
  const { adapter, jobId } = options;
  const requestId = createId("video-status");
  const startTime = Date.now();
  aiEventClient.emit("video:request:started", {
    requestId,
    provider: adapter.name,
    model: adapter.model,
    requestType: "status",
    jobId,
    timestamp: startTime
  });
  const statusResult = await adapter.getVideoStatus(jobId);
  if (statusResult.status === "completed") {
    try {
      const urlResult = await adapter.getVideoUrl(jobId);
      aiEventClient.emit("video:request:completed", {
        requestId,
        provider: adapter.name,
        model: adapter.model,
        requestType: "status",
        jobId,
        status: statusResult.status,
        progress: statusResult.progress,
        url: urlResult.url,
        duration: Date.now() - startTime,
        timestamp: Date.now()
      });
      return {
        status: statusResult.status,
        progress: statusResult.progress,
        url: urlResult.url
      };
    } catch (error) {
      aiEventClient.emit("video:request:completed", {
        requestId,
        provider: adapter.name,
        model: adapter.model,
        requestType: "status",
        jobId,
        status: statusResult.status,
        progress: statusResult.progress,
        error: error instanceof Error ? error.message : "Failed to get video URL",
        duration: Date.now() - startTime,
        timestamp: Date.now()
      });
      return {
        status: statusResult.status,
        progress: statusResult.progress,
        error: error instanceof Error ? error.message : "Failed to get video URL"
      };
    }
  }
  aiEventClient.emit("video:request:completed", {
    requestId,
    provider: adapter.name,
    model: adapter.model,
    requestType: "status",
    jobId,
    status: statusResult.status,
    progress: statusResult.progress,
    error: statusResult.error,
    duration: Date.now() - startTime,
    timestamp: Date.now()
  });
  return {
    status: statusResult.status,
    progress: statusResult.progress,
    error: statusResult.error
  };
}
function createVideoOptions(options) {
  return options;
}
export {
  createVideoOptions,
  generateVideo,
  getVideoJobStatus,
  kind
};
