// @tanstack/ai-openai/dist/esm/adapters/image.js
import { BaseImageAdapter } from "@tanstack/ai/adapters";
import { createOpenAIClient, generateId, getOpenAIApiKeyFromEnv } from "../utils/client.js";
import { validatePrompt, validateImageSize, validateNumberOfImages } from "../image/image-provider-options.js";
class OpenAIImageAdapter extends BaseImageAdapter {
  constructor(config, model) {
    super({}, model);
    this.kind = "image";
    this.name = "openai";
    this.client = createOpenAIClient(config);
  }
  async generateImages(options) {
    const { model, prompt, numberOfImages, size } = options;
    validatePrompt({ prompt, model });
    validateImageSize(model, size);
    validateNumberOfImages(model, numberOfImages);
    const request = this.buildRequest(options);
    const response = await this.client.images.generate({
      ...request,
      stream: false
    });
    return this.transformResponse(model, response);
  }
  buildRequest(options) {
    const { model, prompt, numberOfImages, size, modelOptions } = options;
    return {
      model,
      prompt,
      n: numberOfImages ?? 1,
      size,
      ...modelOptions
    };
  }
  transformResponse(model, response) {
    const images = (response.data ?? []).map((item) => ({
      b64Json: item.b64_json,
      url: item.url,
      revisedPrompt: item.revised_prompt
    }));
    return {
      id: generateId(this.name),
      model,
      images,
      usage: response.usage ? {
        inputTokens: response.usage.input_tokens,
        outputTokens: response.usage.output_tokens,
        totalTokens: response.usage.total_tokens
      } : void 0
    };
  }
}
function createOpenaiImage(model, apiKey, config) {
  return new OpenAIImageAdapter({ apiKey, ...config }, model);
}
function openaiImage(model, config) {
  const apiKey = getOpenAIApiKeyFromEnv();
  return createOpenaiImage(model, apiKey, config);
}
export {
  OpenAIImageAdapter,
  createOpenaiImage,
  openaiImage
};
