// @tanstack/ai-openai/dist/esm/adapters/summarize.js
import { BaseSummarizeAdapter } from "@tanstack/ai/adapters";
import { getOpenAIApiKeyFromEnv } from "../utils/client.js";
import { OpenAITextAdapter } from "./text.js";
class OpenAISummarizeAdapter extends BaseSummarizeAdapter {
  constructor(config, model) {
    super({}, model);
    this.kind = "summarize";
    this.name = "openai";
    this.textAdapter = new OpenAITextAdapter(config, model);
  }
  async summarize(options) {
    const systemPrompt = this.buildSummarizationPrompt(options);
    let summary = "";
    const id = "";
    let model = options.model;
    let usage = { promptTokens: 0, completionTokens: 0, totalTokens: 0 };
    for await (const chunk of this.textAdapter.chatStream({
      model: options.model,
      messages: [{ role: "user", content: options.text }],
      systemPrompts: [systemPrompt],
      maxTokens: options.maxLength,
      temperature: 0.3
    })) {
      if (chunk.type === "TEXT_MESSAGE_CONTENT") {
        if (chunk.content) {
          summary = chunk.content;
        } else {
          summary += chunk.delta;
        }
        model = chunk.model || model;
      }
      if (chunk.type === "RUN_FINISHED") {
        if (chunk.usage) {
          usage = chunk.usage;
        }
      }
    }
    return { id, model, summary, usage };
  }
  async *summarizeStream(options) {
    const systemPrompt = this.buildSummarizationPrompt(options);
    yield* this.textAdapter.chatStream({
      model: options.model,
      messages: [{ role: "user", content: options.text }],
      systemPrompts: [systemPrompt],
      maxTokens: options.maxLength,
      temperature: 0.3
    });
  }
  buildSummarizationPrompt(options) {
    let prompt = "You are a professional summarizer. ";
    switch (options.style) {
      case "bullet-points":
        prompt += "Provide a summary in bullet point format. ";
        break;
      case "paragraph":
        prompt += "Provide a summary in paragraph format. ";
        break;
      case "concise":
        prompt += "Provide a very concise summary in 1-2 sentences. ";
        break;
      default:
        prompt += "Provide a clear and concise summary. ";
    }
    if (options.focus && options.focus.length > 0) {
      prompt += `Focus on the following aspects: ${options.focus.join(", ")}. `;
    }
    if (options.maxLength) {
      prompt += `Keep the summary under ${options.maxLength} tokens. `;
    }
    return prompt;
  }
}
function createOpenaiSummarize(model, apiKey, config) {
  return new OpenAISummarizeAdapter({ apiKey, ...config }, model);
}
function openaiSummarize(model, config) {
  const apiKey = getOpenAIApiKeyFromEnv();
  return createOpenaiSummarize(model, apiKey, config);
}
export {
  OpenAISummarizeAdapter,
  createOpenaiSummarize,
  openaiSummarize
};
