// @tanstack/ai-gemini/dist/esm/adapters/summarize.js
import { FinishReason } from "@google/genai";
import { createGeminiClient, generateId, getGeminiApiKeyFromEnv } from "../utils/client.js";
const GeminiSummarizeModels = [
  "gemini-2.0-flash",
  "gemini-1.5-flash",
  "gemini-1.5-pro",
  "gemini-2.0-flash-lite"
];
class GeminiSummarizeAdapter {
  constructor(config, model) {
    this.kind = "summarize";
    this.name = "gemini";
    this.client = createGeminiClient(config);
    this.model = model;
  }
  async summarize(options) {
    const model = options.model;
    const formatInstructions = this.getFormatInstructions(options.style);
    const lengthInstructions = options.maxLength ? ` Keep the summary under ${options.maxLength} tokens.` : "";
    const systemPrompt = `You are a helpful assistant that summarizes text. ${formatInstructions}${lengthInstructions}`;
    const response = await this.client.models.generateContent({
      model,
      contents: [
        {
          role: "user",
          parts: [
            { text: `Please summarize the following:

${options.text}` }
          ]
        }
      ],
      config: {
        systemInstruction: systemPrompt
      }
    });
    const summary = response.text ?? "";
    const inputTokens = response.usageMetadata?.promptTokenCount ?? 0;
    const outputTokens = response.usageMetadata?.candidatesTokenCount ?? 0;
    return {
      id: generateId("sum"),
      model,
      summary,
      usage: {
        promptTokens: inputTokens,
        completionTokens: outputTokens,
        totalTokens: inputTokens + outputTokens
      }
    };
  }
  async *summarizeStream(options) {
    const model = options.model;
    const id = generateId("sum");
    let accumulatedContent = "";
    let inputTokens = 0;
    let outputTokens = 0;
    const formatInstructions = this.getFormatInstructions(options.style);
    const lengthInstructions = options.maxLength ? ` Keep the summary under ${options.maxLength} words.` : "";
    const systemPrompt = `You are a helpful assistant that summarizes text. ${formatInstructions}${lengthInstructions}`;
    const result = await this.client.models.generateContentStream({
      model,
      contents: [
        {
          role: "user",
          parts: [
            { text: `Please summarize the following:

${options.text}` }
          ]
        }
      ],
      config: {
        systemInstruction: systemPrompt
      }
    });
    for await (const chunk of result) {
      if (chunk.usageMetadata) {
        inputTokens = chunk.usageMetadata.promptTokenCount ?? inputTokens;
        outputTokens = chunk.usageMetadata.candidatesTokenCount ?? outputTokens;
      }
      if (chunk.candidates?.[0]?.content?.parts) {
        for (const part of chunk.candidates[0].content.parts) {
          if (part.text) {
            accumulatedContent += part.text;
            yield {
              type: "TEXT_MESSAGE_CONTENT",
              messageId: id,
              model,
              timestamp: Date.now(),
              delta: part.text,
              content: accumulatedContent
            };
          }
        }
      }
      const finishReason = chunk.candidates?.[0]?.finishReason;
      if (finishReason === FinishReason.STOP || finishReason === FinishReason.MAX_TOKENS || finishReason === FinishReason.SAFETY) {
        yield {
          type: "RUN_FINISHED",
          runId: id,
          model,
          timestamp: Date.now(),
          finishReason: finishReason === FinishReason.STOP ? "stop" : finishReason === FinishReason.MAX_TOKENS ? "length" : "content_filter",
          usage: {
            promptTokens: inputTokens,
            completionTokens: outputTokens,
            totalTokens: inputTokens + outputTokens
          }
        };
      }
    }
  }
  getFormatInstructions(style) {
    switch (style) {
      case "bullet-points":
        return "Provide the summary as bullet points.";
      case "concise":
        return "Provide a very brief one or two sentence summary.";
      case "paragraph":
      default:
        return "Provide the summary in paragraph form.";
    }
  }
}
function createGeminiSummarize(apiKey, model, config) {
  return new GeminiSummarizeAdapter({ ...config, apiKey }, model);
}
function geminiSummarize(model, config) {
  const apiKey = getGeminiApiKeyFromEnv();
  return new GeminiSummarizeAdapter({ ...config, apiKey }, model);
}
export {
  GeminiSummarizeAdapter,
  GeminiSummarizeModels,
  createGeminiSummarize,
  geminiSummarize
};
