// @tanstack/ai-gemini/dist/esm/adapters/tts.js
import { BaseTTSAdapter } from "@tanstack/ai/adapters";
import { createGeminiClient, generateId, getGeminiApiKeyFromEnv } from "../utils/client.js";
class GeminiTTSAdapter extends BaseTTSAdapter {
  constructor(config, model) {
    super(config, model);
    this.name = "gemini";
    this.client = createGeminiClient(config);
  }
  /**
   * Generate speech from text using Gemini's TTS model.
   *
   * @experimental This implementation is experimental and may change.
   * @see https://ai.google.dev/gemini-api/docs/speech-generation
   */
  async generateSpeech(options) {
    const { model, text, modelOptions } = options;
    const voiceConfig = modelOptions?.voiceConfig || {
      prebuiltVoiceConfig: {
        voiceName: "Kore"
      }
    };
    const response = await this.client.models.generateContent({
      model,
      contents: [
        {
          role: "user",
          parts: [{ text }]
        }
      ],
      config: {
        responseModalities: ["AUDIO"],
        speechConfig: {
          voiceConfig,
          ...modelOptions?.languageCode && {
            languageCode: modelOptions.languageCode
          }
        }
      },
      ...modelOptions?.systemInstruction && {
        systemInstruction: modelOptions.systemInstruction
      }
    });
    const candidate = response.candidates?.[0];
    const parts = candidate?.content?.parts;
    if (!parts || parts.length === 0) {
      throw new Error("No audio output received from Gemini TTS");
    }
    const audioPart = parts.find(
      (part) => part.inlineData?.mimeType?.startsWith("audio/")
    );
    if (!audioPart || !audioPart.inlineData || !audioPart.inlineData.data) {
      throw new Error("No audio data in Gemini TTS response");
    }
    const audioBase64 = audioPart.inlineData.data;
    const mimeType = audioPart.inlineData.mimeType || "audio/wav";
    const format = mimeType.split("/")[1] || "wav";
    return {
      id: generateId(this.name),
      model,
      audio: audioBase64,
      format,
      contentType: mimeType
    };
  }
}
function createGeminiSpeech(model, apiKey, config) {
  return new GeminiTTSAdapter({ apiKey, ...config }, model);
}
function geminiSpeech(model, config) {
  const apiKey = getGeminiApiKeyFromEnv();
  return createGeminiSpeech(model, apiKey, config);
}
export {
  GeminiTTSAdapter,
  createGeminiSpeech,
  geminiSpeech
};
