// @tanstack/ai-gemini/dist/esm/tools/tool-converter.js
import { convertCodeExecutionToolToAdapterFormat } from "./code-execution-tool.js";
import { convertComputerUseToolToAdapterFormat } from "./computer-use-tool.js";
import { convertFileSearchToolToAdapterFormat } from "./file-search-tool.js";
import { convertGoogleMapsToolToAdapterFormat } from "./google-maps-tool.js";
import { convertGoogleSearchRetrievalToolToAdapterFormat } from "./google-search-retriveal-tool.js";
import { convertGoogleSearchToolToAdapterFormat } from "./google-search-tool.js";
import { convertUrlContextToolToAdapterFormat } from "./url-context-tool.js";
function convertToolsToProviderFormat(tools) {
  if (!tools || tools.length === 0) {
    return [];
  }
  const result = [];
  const functionDeclarations = [];
  for (const tool of tools) {
    const name = tool.name;
    switch (name) {
      case "code_execution":
        result.push(convertCodeExecutionToolToAdapterFormat());
        break;
      case "computer_use":
        result.push(convertComputerUseToolToAdapterFormat(tool));
        break;
      case "file_search":
        result.push(convertFileSearchToolToAdapterFormat(tool));
        break;
      case "google_maps":
        result.push(convertGoogleMapsToolToAdapterFormat(tool));
        break;
      case "google_search_retrieval":
        result.push(convertGoogleSearchRetrievalToolToAdapterFormat(tool));
        break;
      case "google_search":
        result.push(convertGoogleSearchToolToAdapterFormat(tool));
        break;
      case "url_context":
        result.push(convertUrlContextToolToAdapterFormat());
        break;
      default:
        if (!tool.description) {
          throw new Error(
            `Tool ${tool.name} requires a description for Gemini adapter`
          );
        }
        functionDeclarations.push({
          name: tool.name,
          description: tool.description,
          parameters: tool.inputSchema ?? {
            type: "object",
            properties: {},
            required: []
          }
        });
        break;
    }
  }
  if (functionDeclarations.length > 0) {
    result.push({
      functionDeclarations
    });
  }
  return result;
}
export {
  convertToolsToProviderFormat
};
