// @tanstack/ai-openai/dist/esm/index.js
import { OpenAITextAdapter, createOpenaiChat, openaiText } from "./adapters/text.js";
import { OpenAISummarizeAdapter, createOpenaiSummarize, openaiSummarize } from "./adapters/summarize.js";
import { OpenAIImageAdapter, createOpenaiImage, openaiImage } from "./adapters/image.js";
import { OpenAIVideoAdapter, createOpenaiVideo, openaiVideo } from "./adapters/video.js";
import { OpenAITTSAdapter, createOpenaiSpeech, openaiSpeech } from "./adapters/tts.js";
import { OpenAITranscriptionAdapter, createOpenaiTranscription, openaiTranscription } from "./adapters/transcription.js";
import { OPENAI_CHAT_MODELS, OPENAI_IMAGE_MODELS, OPENAI_TRANSCRIPTION_MODELS, OPENAI_TTS_MODELS, OPENAI_VIDEO_MODELS } from "./model-meta.js";
export {
  OPENAI_CHAT_MODELS,
  OPENAI_IMAGE_MODELS,
  OPENAI_TRANSCRIPTION_MODELS,
  OPENAI_TTS_MODELS,
  OPENAI_VIDEO_MODELS,
  OpenAIImageAdapter,
  OpenAISummarizeAdapter,
  OpenAITTSAdapter,
  OpenAITextAdapter,
  OpenAITranscriptionAdapter,
  OpenAIVideoAdapter,
  createOpenaiChat,
  createOpenaiImage,
  createOpenaiSpeech,
  createOpenaiSummarize,
  createOpenaiTranscription,
  createOpenaiVideo,
  openaiImage,
  openaiSpeech,
  openaiSummarize,
  openaiText,
  openaiTranscription,
  openaiVideo
};
