// @tanstack/ai-openai/dist/esm/model-meta.js
const GPT5_2 = {
  name: "gpt-5.2"
};
const GPT5_2_PRO = {
  name: "gpt-5.2-pro"
};
const GPT5_2_CHAT = {
  name: "gpt-5.2-chat-latest"
};
const GPT5_1 = {
  name: "gpt-5.1"
};
const GPT5_1_CODEX = {
  name: "gpt-5.1-codex"
};
const GPT5 = {
  name: "gpt-5"
};
const GPT5_MINI = {
  name: "gpt-5-mini"
};
const GPT5_NANO = {
  name: "gpt-5-nano"
};
const GPT5_PRO = {
  name: "gpt-5-pro"
};
const GPT5_CODEX = {
  name: "gpt-5-codex"
};
const SORA2 = {
  name: "sora-2"
};
const SORA2_PRO = {
  name: "sora-2-pro"
};
const GPT_IMAGE_1 = {
  name: "gpt-image-1"
};
const GPT_IMAGE_1_MINI = {
  name: "gpt-image-1-mini"
};
const O3_DEEP_RESEARCH = {
  name: "o3-deep-research"
};
const O4_MINI_DEEP_RESEARCH = {
  name: "o4-mini-deep-research"
};
const O3_PRO = {
  name: "o3-pro"
};
const GPT_AUDIO = {
  name: "gpt-audio"
};
const GPT_AUDIO_MINI = {
  name: "gpt-audio-mini"
};
const O3 = {
  name: "o3"
};
const O4_MINI = {
  name: "o4-mini"
};
const GPT4_1 = {
  name: "gpt-4.1"
};
const GPT4_1_MINI = {
  name: "gpt-4.1-mini"
};
const GPT4_1_NANO = {
  name: "gpt-4.1-nano"
};
const O1_PRO = {
  name: "o1-pro"
};
const COMPUTER_USE_PREVIEW = {
  name: "computer-use-preview"
};
const GPT_4O_MINI_SEARCH_PREVIEW = {
  name: "gpt-4o-mini-search-preview"
};
const GPT_4O_SEARCH_PREVIEW = {
  name: "gpt-4o-search-preview"
};
const O3_MINI = {
  name: "o3-mini"
};
const GPT_4O_MINI_AUDIO = {
  name: "gpt-4o-mini-audio"
};
const O1 = {
  name: "o1"
};
const GPT_4O = {
  name: "gpt-4o"
};
const GPT_4O_AUDIO = {
  name: "gpt-4o-audio"
};
const GPT_4O_MINI = {
  name: "gpt-4o-mini"
};
const GPT_4_TURBO = {
  name: "gpt-4-turbo"
};
const CHATGPT_40 = {
  name: "chatgpt-4o-latest"
};
const GPT_5_1_CODEX_MINI = {
  name: "gpt-5.1-codex-mini"
};
const CODEX_MINI_LATEST = {
  name: "codex-mini-latest"
};
const DALL_E_2 = {
  name: "dall-e-2"
};
const DALL_E_3 = {
  name: "dall-e-3"
};
const GPT_3_5_TURBO = {
  name: "gpt-3.5-turbo"
};
const GPT_4 = {
  name: "gpt-4"
};
const GPT_5_1_CHAT = {
  name: "gpt-5.1-chat-latest"
};
const GPT_5_CHAT = {
  name: "gpt-5-chat-latest"
};
const OPENAI_CHAT_MODELS = [
  // Frontier models
  GPT5_2.name,
  GPT5_2_PRO.name,
  GPT5_2_CHAT.name,
  GPT5_1.name,
  GPT5_1_CODEX.name,
  GPT5.name,
  GPT5_MINI.name,
  GPT5_NANO.name,
  GPT5_PRO.name,
  GPT5_CODEX.name,
  // Reasoning models
  O3.name,
  O3_PRO.name,
  O3_MINI.name,
  O4_MINI.name,
  O3_DEEP_RESEARCH.name,
  O4_MINI_DEEP_RESEARCH.name,
  // GPT-4 series
  GPT4_1.name,
  GPT4_1_MINI.name,
  GPT4_1_NANO.name,
  GPT_4.name,
  GPT_4_TURBO.name,
  GPT_4O.name,
  GPT_4O_MINI.name,
  // GPT-3.5
  GPT_3_5_TURBO.name,
  // Audio-enabled chat models
  GPT_AUDIO.name,
  GPT_AUDIO_MINI.name,
  GPT_4O_AUDIO.name,
  GPT_4O_MINI_AUDIO.name,
  // ChatGPT models
  GPT_5_1_CHAT.name,
  GPT_5_CHAT.name,
  CHATGPT_40.name,
  // Specialized
  GPT_5_1_CODEX_MINI.name,
  CODEX_MINI_LATEST.name,
  // Preview models
  GPT_4O_SEARCH_PREVIEW.name,
  GPT_4O_MINI_SEARCH_PREVIEW.name,
  COMPUTER_USE_PREVIEW.name,
  // Legacy reasoning
  O1.name,
  O1_PRO.name
];
const OPENAI_IMAGE_MODELS = [
  GPT_IMAGE_1.name,
  GPT_IMAGE_1_MINI.name,
  DALL_E_3.name,
  DALL_E_2.name
];
const OPENAI_VIDEO_MODELS = [SORA2.name, SORA2_PRO.name];
const OPENAI_TTS_MODELS = [
  "tts-1",
  "tts-1-hd",
  "gpt-4o-audio-preview"
];
const OPENAI_TRANSCRIPTION_MODELS = [
  "whisper-1",
  "gpt-4o-transcribe",
  "gpt-4o-mini-transcribe",
  "gpt-4o-transcribe-diarize"
];
export {
  OPENAI_CHAT_MODELS,
  OPENAI_IMAGE_MODELS,
  OPENAI_TRANSCRIPTION_MODELS,
  OPENAI_TTS_MODELS,
  OPENAI_VIDEO_MODELS
};
